crates/uv-resolver/src/resolver/mod.rs

```diff
@@ -355,11 +375,285 @@ impl<'a, Provider: ResolverProvider, InstalledPackages: InstalledPackagesProvide
+                    &self.preferences,
+                    self.installed_packages,
+                    &self.exclusions,
+                ) else {
+                    return Ok(None);
+                };
+
+                // If there is not a compatible distribution, short-circuit.
+                let Some(dist) = candidate.compatible() else {
+                    return Ok(None);
+                };
+
+                // Emit a request to fetch the metadata for this version.
+                if self.index.distributions.register(candidate.version_id()) {
+                    let dist = dist.for_resolution().to_owned();
+
+                    let response = match dist {
+                        ResolvedDist::Installable(dist) => {
+                            let metadata = self
+                                .provider
+                                .get_or_build_wheel_metadata(&dist)
+                                .boxed_local()
+                                .await
+                                .map_err(|err| match dist.clone() {
+                                    Dist::Built(built_dist @ BuiltDist::Path(_)) => {
+                                        ResolveError::Read(Box::new(built_dist), err)
+                                    }
+                                    Dist::Source(source_dist @ SourceDist::Path(_)) => {
+                                        ResolveError::Build(Box::new(source_dist), err)
+                                    }
+                                    Dist::Source(source_dist @ SourceDist::Directory(_)) => {
+                                        ResolveError::Build(Box::new(source_dist), err)
+                                    }
+                                    Dist::Built(built_dist) => {
+                                        ResolveError::Fetch(Box::new(built_dist), err)
+                                    }
+                                    Dist::Source(source_dist) => {
+                                        ResolveError::FetchAndBuild(Box::new(source_dist), err)
+                                    }
+                                })?;
+                            Response::Dist { dist, metadata }
+                        }
+                        ResolvedDist::Installed(dist) => {
+                            let metadata = dist.metadata().map_err(|err| {
+                                ResolveError::ReadInstalled(Box::new(dist.clone()), err)
+                            })?;
+                            Response::Installed { dist, metadata }
+                        }
+                    };
+
+                    Ok(Some(response))
+                } else {
+                    Ok(None)
+                }
+            }
+        }
+    }
+
+    fn on_complete(&self) {
+        if let Some(reporter) = self.reporter.as_ref() {
+            reporter.on_complete();
+        }
+    }
+}
+
+struct Solver<'a, InstalledPackages: InstalledPackagesProvider> {
```

Comment:
What is the difference/relationship between this type and `Resolver` and `SolverState`? I'm bouncing between them trying to find a difference. I see `Resolver` is generic over a `Provider` where this is not. Some code comments explaining what this type is and why it exists would help I think. (I think the code comments should, at minimum, answer the question of why there is duplication here.)